File: application/service/blocking/src/main/java/org/occurrent/application/service/blocking/GenericApplicationService.java
package org.occurrent.application.service.blocking;

import io.cloudevents.CloudEvent;
import org.occurrent.eventstore.api.blocking.EventStore;
import org.occurrent.eventstore.api.blocking.EventStream;

import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Stream;

public class GenericApplicationService<T> {

    private final EventStore eventStore;
    private final CloudEventConverter<T> cloudEventConverter;

    public GenericApplicationService(EventStore eventStore, CloudEventConverter<T> cloudEventConverter) {
        this.eventStore = eventStore;
        this.cloudEventConverter = cloudEventConverter;
    }

    public void execute(UUID streamId, Function<Stream<T>, Stream<T>> functionThatCallsDomainModel) {
        Objects.requireNonNull(streamId, "Stream id cannot be null");
        execute(streamId.toString(), functionThatCallsDomainModel);
    }

    /**
     *
     * @param streamId
     * @param functionThatCallsDomainModel
     */
    public void execute(String streamId, Function<Stream<T>, Stream<T>> functionThatCallsDomainModel) {
        Objects.requireNonNull(streamId, "Stream id cannot be null");
        Objects.requireNonNull(functionThatCallsDomainModel, "Function that calls domain model cannot be null");
        // Read all events from the event store for a particular stream
        EventStream<CloudEvent> eventStream = eventStore.read(streamId);

        // Convert the cloud events into domain events
        Stream<T> eventsInStream = eventStream.events().map(cloudEventConverter::toDomainEvent);

        // Call a pure function from the domain model which returns a Stream of events
        Stream<CloudEvent> newEvents = functionThatCallsDomainModel.apply(eventsInStream).map(cloudEventConverter::toCloudEvent);

        // Write the new events to the event store
        eventStore.write(streamId, eventStream.version(), newEvents);
    }
}

File: example/domain/word-guessing-game/mongodb/spring/blocking/src/main/kotlin/org/occurrent/example/domain/wordguessinggame/mongodb/spring/blocking/infrastructure/GenericApplicationService.kt
package org.occurrent.example.domain.wordguessinggame.mongodb.spring.blocking.infrastructure

import io.cloudevents.CloudEvent
import org.occurrent.eventstore.api.WriteConditionNotFulfilledException
import org.occurrent.eventstore.api.blocking.EventStore
import org.occurrent.eventstore.api.blocking.EventStream
import org.occurrent.example.domain.wordguessinggame.event.DomainEvent
import org.springframework.retry.annotation.Backoff
import org.springframework.retry.annotation.Retryable
import org.springframework.stereotype.Service
import java.util.*
import kotlin.streams.asSequence
import kotlin.streams.asStream

@Service
class GenericApplicationService constructor(private val eventStore: EventStore,
                                            private val cloudEventConverter: CloudEventConverter) {

    @Retryable(include = [WriteConditionNotFulfilledException::class], maxAttempts = 5, backoff = Backoff(delay = 100, multiplier = 2.0, maxDelay = 1000))
    fun execute(streamId: UUID, functionThatCallsDomainModel: (Sequence<DomainEvent>) -> Sequence<DomainEvent>) {
        // Read all events from the event store for a particular stream
        val eventStream: EventStream<CloudEvent> = eventStore.read(streamId.toString())
        // Convert the cloud events into domain events
        val domainEventsInStream: Sequence<DomainEvent> = eventStream.events().map(cloudEventConverter::toDomainEvent).asSequence()

        // Call a pure function from the domain model which returns a sequence of domain events
        val newDomainEvents = functionThatCallsDomainModel(domainEventsInStream)

        // Convert domain events to cloud events and write them to the event store
        eventStore.write(streamId.toString(), eventStream.version(), newDomainEvents.map(cloudEventConverter::toCloudEvent).asStream())
    }
}